File: 프로그래머스/2/12953. N개의 최소공배수/N개의 최소공배수.py
from collections import deque

def gcd(a, b):
    while b > 0:
        a, b = b, a % b
    return a

def lcm(a, b):
    return (a * b) // gcd(a, b)

def solution(arr):
    q = deque(arr)
    while len(q) != 1:
        for i in range(len(q) - 1):
            a, b = q.popleft(), q.popleft()
            q.append(lcm(a, b))
    return q[0]

File: 프로그래머스/2/12953. N개의 최소공배수/N개의 최소공배수.java
import java.util.*;

class Solution {
    public int solution(int[] arr) {
        Deque<Integer> dq = new ArrayDeque<>();
        for (int a : arr) {
            dq.push(a);
        }
        
        while (dq.size() != 1) {
            int size = dq.size();
            for (int i = 0; i < size - 1; i++) {
                int a = dq.pop();
                int b = dq.pop();
                dq.push(lcm(a, b));
            }
        }
        return dq.peek();
    }
    
    public int gcd(int a, int b) {
        int temp = 0;
        while (b > 0) {
            temp = a;
            a = b; b = temp % b;
        }
        return a;
    }
    
    public int lcm(int a, int b) {
        return (int) (a * b) / gcd(a, b);
    }
}